### PR TITLE
Refactor score unwinding

### PR DESCRIPTION
**Pull request by jeetee**
Resolves: issues with Jump/repeat combinations: [#287447](https://musescore.org/en/node/287447), [#299320](https://musescore.org/en/node/299320), [#300362](https://musescore.org/en/node/300362), [#304795](https://musescore.org/en/node/304795)
[#284887: crash on playback end in Debug build](https://musescore.org/en/node/284887)

Refactored unwinding to mimic "real" interpretation chain by a human over the previous copy-and-paste-rewind logic for jumps.

Also tracks section break pauses on RepeatSegments (but not used to fix related issue [#32696](https://musescore.org/en/node/32696) yet)

- [x] I signed [CLA](https://musescore.org/en/cla)
- [x] I made sure the code in the PR follows [the coding rules](https://github.com/musescore/Documentation/blob/master/WorkflowAndGuidelines/CodeGuidelines.md)
- [x] I made sure the code compiles on my machine
- [x] I made sure there are no unnecessary changes in the code
- [x] I made sure the title of the PR reflects the core meaning of the issue you are solving
- [x] I made sure the commit message(s) contain a description and answer the question "Why do those changes fix that particular issue?" or "Why are those changes really necessary as improvements?"
- [x] I made sure the commit message title starts with "fix #424242:" if there is a related issue
- [x] I created the test (mtest, vtest, script test) to verify the changes I made

### CODE REVIEW

File: libmscore/rendermidi.cpp [hunk 1]
```diff
@@ -748,7 +748,7 @@ void MidiRenderer::collectMeasureEventsDefault(EventMap* events, Measure* m, con
 //    redirects to the correct function based on the passed method
 //---------------------------------------------------------
 
-void MidiRenderer::collectMeasureEvents(EventMap* events, Measure* m, const StaffContext& sctx, int tickOffset)
+void MidiRenderer::collectMeasureEvents(EventMap* events, Measure const * m, const StaffContext& sctx, int tickOffset)
```
Comment (jthistle) [inline libmscore/rendermidi.cpp:751]: Correct me if I'm wrong, but don't we prefer `const Measure*` over `Measure const *`? Or maybe that's just an inconsequential thing, I don't know :)

**Comment by anatoly-os** [comment]
You really want to put these changes to 3.5RC, don't you? :) I suggest postponing to 3.5.1. What do you think?

File: libmscore/rendermidi.cpp (hunk 1 above)
Comment (jeetee) [reply libmscore/rendermidi.cpp:751]: I believe the order wasn't specified in the "old" coding rules in effect on 3.x

The new ones indeed write `const T*` as a preference; but I'm still fighting that rule from a consistency pov, One should read type declarations from right to left and that would mean language natural that `T const *` makes more sense.
Anyhow.. let's not get into that discussion here.

If it must be changed to abide by coding rules, then I'll make that change; if I don't *have* to, then I won't ;-)

**Comment by jeetee** [comment]
3.5.1 is fine too @anatoly-os 
I just didn't want it to wait for MS4 as the related bugs (especially D.S. at section break misfunctioning) does come up regularly (about once a month) in the forums.

And of course my code is so perfect that any PR I make warrants its own release just for the benefit of our users ;)


So yeah, preferably before 4 will do

File: libmscore/rendermidi.cpp (hunk 1 above)
Comment (jthistle) [reply libmscore/rendermidi.cpp:751]: Yeah, please don't feel you have to just because I've piped up ;)

Comment (Marr11317) [reply libmscore/rendermidi.cpp:751]: wow, I just realised that jeetee and jthistle were not the same person after all 😆 
All that time I thought both were a short for johan temmerman, pointing to the same person...

Comment (jeetee) [reply libmscore/rendermidi.cpp:751]: @Marr11317 Who says we're not the same person? 😇 
I'm that talkative one that posts in the forums and occasionally fixes something.
James is the one that quietly fixes a bunch of stuff and occasionally speaks up.

Comment (RobFog) [reply libmscore/rendermidi.cpp:751]: And I believe there's a certain difference in age. 😉

Comment (Jojo-Schmitz) [reply libmscore/rendermidi.cpp:751]: Not a big one

**Comment by jeetee** [comment]
3.5.1 is even better.
https://musescore.org/en/node/307231 exposed a bug with two volta's and a double repeat inbetween them.

|:    |    volta1 :||: volta2 |      :|
wrongly includes the last measure of volta1 on the 2nd repeat

**Comment by jeetee** [comment]
Rebased and added test::repeat57 to cover https://musescore.org/en/node/274690#comment-1013746

File: libmscore/score.cpp [hunk 2]
```diff
@@ -3808,7 +3808,10 @@ void MasterScore::setPos(POS pos, Fraction tick)
       {
       if (tick < Fraction(0,1))
             tick = Fraction(0,1);
-      Q_ASSERT(tick <= lastMeasure()->endTick());
+      if (tick > lastMeasure()->endTick()) {
+            // End Reverb may last longer than written notation, but cursor position should not
+            tick = lastMeasure()->endTick();
+            }
```
Comment (Jojo-Schmitz) [inline libmscore/score.cpp:3814]: See #6437

**Comment by anatoly-os** [comment]
That was not the easiest one, but see 87a89319ecde265b56787a4377b6292d3b495156 :)

**Comment by jeetee** [comment]
Easy is a relative term ;-)
Thanks for the additional effort to port it to master!

**Pull request merged**